File: 2020/day02/day02.py
import sys

fname = '../input/day02/input.txt'

part2 = False
if len(sys.argv) > 1 and sys.argv[1] == '2':
    part2 = True

valid = 0

with open(fname) as fh:
    for line in fh.readlines():
        parts = line.split(':')
        passw = parts[1].strip()
        parts = parts[0].split(' ')
        match = parts[1]
        parts = parts[0].split('-')
        n1 = int(parts[0])
        n2 = int(parts[1])
        if part2:
            if (passw[n1-1] == match and not passw[n2-1] == match) or \
                    (not passw[n1-1] == match and passw[n2-1] == match):
                valid = valid + 1
        else:
            vc = 0
            for i in range(0, len(passw)):
                if passw[i] == match:
                    vc = vc + 1
            if vc >= n1 and vc <= n2:
                valid = valid + 1
    print(valid)

File: 2020/java/aoc2020/src/main/java/org/f5n/aoc2020/days/Day02.java
package org.f5n.aoc2020.days;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import org.f5n.aoc2020.utils.*;


public class Day02 extends Day {
	protected List<String> input;
	protected int valid = 0;

	public Day02(List<String> input) {
		this.input = input;
	}

	protected Result part1(Input limit) {
		valid = 0;
		run(1);
		return new IntResult(valid);
	}

	protected Result part2(Input limit) {
		valid = 0;
		run(2);
		return new IntResult(valid);
	}

	private void run(int part) {
		for (String s : input) {
			String[] parts = s.split(":");
			String passw = parts[1].strip();
			parts = parts[0].split(" ");
			char match = parts[1].charAt(0);
			parts = parts[0].split("-");
			int n1 = Integer.parseInt(parts[0]);
			int n2 = Integer.parseInt(parts[1]);

			if (part == 1) {
				calc1(passw, match, n1, n2);
			} else {
				calc2(passw, match, n1, n2);
			}
		}
	}

	private void calc1(String passw, char match, int n1, int n2) {
		int vc = 0;
		for (int i = 0; i<passw.length(); ++i) {
			if (passw.charAt(i) == match) {
				++vc;
			}
		}
		if (vc >= n1 && vc <= n2) {
			++valid;
		}
	}

	private void calc2(String passw, char match, int n1, int n2) {
		if ((passw.charAt(n1-1) == match && !(passw.charAt(n2-1) == match)) ||
				(!(passw.charAt(n1-1) == match) && passw.charAt(n2-1) == match)) {
			++valid;
		}
	}
}